driver/src/contracts/batched_auction_data_reader.rs

```diff
@@ -0,0 +1,255 @@
+    pub orders: Vec<Order>,
+    /// The total number of orders per user.
+    pub user_order_counts: HashMap<H160, u16>,
+    /// Used when reading data from the smart contract in batches.
+    pub pagination: Pagination,
+}
+
+/// Data for the next call to the smart contract's `getEncodedUsersPaginated`
+/// function (which should be named `getEncodedOrdersPaginated`).
+pub struct Pagination {
+    pub previous_page_user: H160,
+    pub previous_page_user_offset: u64,
+}
+
+impl BatchedAuctionDataReader {
+    pub fn new(index: U256) -> BatchedAuctionDataReader {
+        BatchedAuctionDataReader {
+            index,
+            account_state: AccountState::default(),
+            orders: Vec::new(),
+            user_order_counts: HashMap::new(),
+            pagination: Pagination {
+                previous_page_user: H160::zero(),
+                previous_page_user_offset: 0,
+            },
+        }
+    }
+
+    /// Applies one batch of data.
+    ///
+    /// A batch can come from `getEncodedUsersPaginated` or `getEncodedOrders`.
+    /// In the latter case there is only one batch.
+    ///
+    /// Returns the number of orders in the data.
+    ///
+    /// Panics if `packed_auction_bytes` is not a multiple of
+    /// `AUCTION_ELEMENT_WIDTH`.
+    pub fn apply_batch(&mut self, packed_auction_bytes: &[u8]) -> u64 {
+        let previous_order_count = self.orders.len();
+        self.apply_auction_data(&packed_auction_bytes);
+        let number_of_added_orders = (self.orders.len() - previous_order_count) as u64;
+        if number_of_added_orders == 0 {
+            return 0;
+        }
+        let last_order_user = self
+            .orders
+            .last()
+            .expect("have at least one order")
+            .account_id;
+        if self.pagination.previous_page_user == last_order_user {
+            self.pagination.previous_page_user_offset += number_of_added_orders;
+        } else {
+            self.pagination.previous_page_user = last_order_user;
+            self.pagination.previous_page_user_offset = number_of_added_orders;
+        }
+        number_of_added_orders
+    }
+
+    fn apply_auction_data(&mut self, packed_auction_bytes: &[u8]) {
+        let auction_elements = self.parse_auction_elements(packed_auction_bytes);
+        self.apply_auction_elements_to_account_state(auction_elements.iter());
+        self.orders.extend(
+            auction_elements
+                .into_iter()
+                .map(|auction_element| auction_element.order),
+        );
+    }
+
+    fn apply_auction_elements_to_account_state<'a, Iter>(&mut self, auction_elements: Iter)
+    where
+        Iter: Iterator<Item = &'a StableXAuctionElement>,
+    {
+        for element in auction_elements.into_iter() {
+            self.account_state.modify_balance(
+                element.order.account_id,
+                element.order.sell_token,
+                |x| *x = element.sell_token_balance,
```

Comment:
I know this wasn't there before, but I'm thinking we should maybe panic here if the value is set and not the same as the new balance. That seems like an invalid state. What do you think?

Reply:
In the paginated version it really could change if  the state of the smart contract changes between different retrievals. Not sure how exceptional that is and if it can be handled  gracefully.

Reply:
Maybe the reader can read for a specific block? Calls can be done on a specific block number which will guarantee the data will not change during retrieval. Because new orders are only valid for the next batch, it would be OK if we miss them.

Reply:
Good idea. I have implemented this.

Reply:
So there is still the question whether we should check and panic. On the one hand if the output of the smart contract cannot be trusted then we cannot expect anything to work correctly here. On the other hand it is a simple check...

Reply:
Technically, the sell_token_balance can only change when also the auctionId changed (every 5 minutes). 
Since we query those orders at the beginning of a batch this shouldn't really happen or otherwise the batch we are looking to process can no longer be solved anyways. However, it can happen if we start the service just before a batch is about to end (we should probably not attempt to process batches if they are already 4 minutes or so in).

Thus I think it's fine to panic as it can be considered invalid state.

Reply:
If we always query with the same block number, then all queries should agree since the batch index will stay the same. Or am I misunderstanding how that works?

Reply:
You are right, my comment was assuming that we wouldn't pin the block number, but that's even better.